airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/debezium/DebeziumComponent.kt

```diff
@@ -0,0 +1,389 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+package io.airbyte.cdk.components.debezium
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.NullNode
+import io.debezium.relational.history.HistoryRecord
+import java.time.Duration
+import java.util.*
+import kotlin.NoSuchElementException
+import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics
+import org.apache.kafka.connect.source.SourceRecord
+
+/** API for collecting CDC records using Debezium. */
+fun interface DebeziumComponent {
+
+    /**
+     * Collect CDC records from the source database as specified by the [input]. This should have
+     * little to no side effects: the state of source database and the [input] determine the output.
+     *
+     * This is intended to be an atomic operation which should be straightforward to mock in
+     * high-level unit tests of a database source. As such its space and time usage are capped
+     * according to paramenters in the [input].
+     *
+     * In production, we need to checkpoint overall progress at regular intervals; the intention
+     * here is to do so after each call to [collect]. The [collectRepeatedly] method can be used to
+     * call [collect] in sequence.
+     */
+    fun collect(input: Input): Output
+
+    /**
+     * Input for [collect].
+     *
+     * This is separated into two parts according to the following principle:
+     * - [config] contains the input parameters which don't change from one successive call to
+     * [collect] to the next;
+     * - [state] contains those which do, these will be the initial content of Debezium's offset and
+     * schema history files.
+     */
+    @JvmRecord data class Input(val config: Config, val state: State)
+
+    /**
+     * [Config] holds the [Input] parameters which don't change from one successive call to
+     * [collect] to the next.
+     *
+     * The [debeziumProperties] property map is used to initialize the Debezium engine.
+     *
+     * The [lsnMapper] interface is used to compare relative CDC progress without knowledge of how
+     * the progress itself is encoded.
+     *
+     * The rest are termination criteria:
+     * - [upperBound] is the targeted upper bound of CDC progress, [collect] stops when it's reached
+     * or exceeded,
+     * - [maxRecords] is the threshold of collected [Record]s at or beyond which [collect] stops,
+     * - [maxRecordBytes] is the threshold of collected JSON [Record] bytes at or beyond which
+     * [collect] stops,
+     * - [maxTime] is the threshold of collected JSON-serialized [Record] bytes at or beyond which
+     * [collect] stops,
+     */
+    data class Config(
+        val debeziumProperties: Properties,
+        val lsnMapper: LsnMapper<*>,
+        val upperBound: State.Offset,
+        val maxRecords: Long,
+        val maxRecordBytes: Long,
+        val maxTime: Duration,
```

Comment:
I'm not sure how much I like that API. why do we need an upper bound? Why do we need maxRecords and maxRecordBytes, or maxTime? I'd rather pass in a queue that will be populated by collect. The duration, maxRecords and maxRecordBytes don't need to be known by the config IMO. It's something that's driven by the CDK, and that should be enforced by blocking the queue or shutting it down

Reply:
I like your idea in principle, for the separation of concerns it brings. Is it achievable, though? The upper bound actually gets checked against heartbeat events which don't make it to this conceptual queue. Arguably, the upper bound is also the most important termination criterion.

I'll think about it! Thanks for the suggestion.

Reply:
I realize that I didn't actually answer your questions:
- We need the upper bound because the sync needs to stop at some point, this is how the connector currently behaves.
- `maxTime` serves two purposes. It's a checkpoint interval, but also is used for detecting when we've caught up with the database and the database is not experiencing any activity. At which point the sync stops, and this is also how the connector currently behaves.
- `maxRecords` is used as a checkpointing trigger
- `maxRecordBytes` is new, because this implementation accumulates records in memory, and also is used as a checkpointing trigger

Reply:
Re. separating the checkpointing and the record production: I'll hack it out next week to see what gives. The thing is, Debezium actually needs to shut down to produce a STATE message that's reliable, because we need to read the offset + schema history files that it maintains. 

So, in effect, in addition to a `void accept(Consumer<Record>)` method the `DebeziumComponent` also needs to expose a `signalShutdown()` method to be called by the `Consumer<Record>` and then also a `State stateAfterShutdown()` which also needs to be plumbed to the consumer or whatever it is that produces `AirbyteMessage`s. I need to think about this properly so as to do this in a way that doesn't devolve into callback hell.

Reply:
To add to this : 
1. Can the maxTime param not be overloaded? We can have a seperate construct for time-based checkpointing, and another for timeout. As a side-note, the time-based checkpointing may not actually be ever used by our code. 
2. Can the checkpointing trigger params be encapsulated into one object? We currently do something similar in our rewritten state iterator.

Reply:
1. In this scheme here there isn't really such a thing as a global timeout, so `maxTime` actually ends up serving as a checkpointing trigger. The main difference is that here, if the checkpointed state is the same as the last, this triggers the global stop.
2. Yes, I agree.